### ISSUE TITLE
loss calculation issue

### ISSUE THREAD

**Issue by iamtekson** [ashokdahal/RiskChanges#39]
Traceback (most recent call last):
  File "/home/sdss/.virtualenvs/sdss/lib/python3.8/site-packages/django/core/handlers/exception.py", line 34, in inner
    response = get_response(request)
  File "/home/sdss/.virtualenvs/sdss/lib/python3.8/site-packages/django/core/handlers/base.py", line 115, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "/home/sdss/.virtualenvs/sdss/lib/python3.8/site-packages/django/core/handlers/base.py", line 113, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/home/sdss/.virtualenvs/sdss/lib/python3.8/site-packages/django/views/decorators/csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "/home/sdss/.virtualenvs/sdss/lib/python3.8/site-packages/rest_framework/viewsets.py", line 114, in view
    return self.dispatch(request, *args, **kwargs)
  File "/home/sdss/.virtualenvs/sdss/lib/python3.8/site-packages/rest_framework/views.py", line 505, in dispatch
    response = self.handle_exception(exc)
  File "/home/sdss/.virtualenvs/sdss/lib/python3.8/site-packages/rest_framework/views.py", line 465, in handle_exception
    self.raise_uncaught_exception(exc)
  File "/home/sdss/.virtualenvs/sdss/lib/python3.8/site-packages/rest_framework/views.py", line 476, in raise_uncaught_exception
    raise exc
  File "/home/sdss/.virtualenvs/sdss/lib/python3.8/site-packages/rest_framework/views.py", line 502, in dispatch
    response = handler(request, *args, **kwargs)
  File "/opt/sdss/django/sdss/loss/views.py", line 100, in create
    loss_calculation(loss_id)
  File "/opt/sdss/django/sdss/loss/tasks.py", line 22, in loss_calculation
    ComputeLoss(CONN_STR, exposure_id, loss_id, loss_type,
  File "/home/sdss/.virtualenvs/sdss/lib/python3.8/site-packages/RiskChanges/ComputeLoss.py", line 76, in ComputeLoss
    exposure = readvector.prepareExposureForLoss(con, exposureid)
  File "/home/sdss/.virtualenvs/sdss/lib/python3.8/site-packages/RiskChanges/RiskChangesOps/readvector.py", line 53, in prepareExposureForLoss
    exposure_all = pd.merge(left=exposuredata, right=eardata, how='left', left_on=[
  File "/home/sdss/.virtualenvs/sdss/lib/python3.8/site-packages/pandas/core/reshape/merge.py", line 74, in merge
    op = _MergeOperation(
  File "/home/sdss/.virtualenvs/sdss/lib/python3.8/site-packages/pandas/core/reshape/merge.py", line 668, in __init__
    ) = self._get_merge_keys()
  File "/home/sdss/.virtualenvs/sdss/lib/python3.8/site-packages/pandas/core/reshape/merge.py", line 1033, in _get_merge_keys
    right_keys.append(right._get_label_or_level_values(rk))
  File "/home/sdss/.virtualenvs/sdss/lib/python3.8/site-packages/pandas/core/generic.py", line 1684, in _get_label_or_level_values
    raise KeyError(key)

**Comment by ashokdahal** [comment]
@iamtekson Could you check whether the EAR metadata have data on the "earPK" field, as well as the exposure data, have a "geom_id" field ? The problem is because the key in data is not available. If true, I will add a try-catch statement to check if that exists on data or not and throw an error.

**Comment by iamtekson** [comment]
Now, I am facing the below error for intensity hazard @ashokdahal,

```py
  File "F:\SDSS\1. code\venv\lib\site-packages\django\core\handlers\exception.py", line 34, in inner
    response = get_response(request)
  File "F:\SDSS\1. code\venv\lib\site-packages\django\core\handlers\base.py", line 115, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "F:\SDSS\1. code\venv\lib\site-packages\django\core\handlers\base.py", line 113, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "F:\SDSS\1. code\venv\lib\site-packages\django\views\decorators\csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "F:\SDSS\1. code\venv\lib\site-packages\rest_framework\viewsets.py", line 114, in view     
    return self.dispatch(request, *args, **kwargs)
  File "F:\SDSS\1. code\venv\lib\site-packages\rest_framework\views.py", line 505, in dispatch    
    response = self.handle_exception(exc)        
  File "F:\SDSS\1. code\venv\lib\site-packages\rest_framework\views.py", line 465, in handle_exception
    self.raise_uncaught_exception(exc)
  File "F:\SDSS\1. code\venv\lib\site-packages\rest_framework\views.py", line 476, in raise_uncaught_exception
    raise exc
  File "F:\SDSS\1. code\venv\lib\site-packages\rest_framework\views.py", line 502, in dispatch    
    response = handler(request, *args, **kwargs) 
  File "F:\SDSS\1. code\django\sdss\loss\views.py", line 100, in create
    loss_calculation(loss_id)
  File "F:\SDSS\1. code\django\sdss\loss\tasks.py", line 26, in loss_calculation
    ComputeLoss(CONN_STR, exposure_id, loss_id, loss_type)
  File "F:\SDSS\1. code\venv\lib\site-packages\RiskChanges\ComputeLoss.py", line 85, in ComputeLoss
    exposure = estimatevulnerability(exposure, haztype, vulnColumn, con)
  File "F:\SDSS\1. code\venv\lib\site-packages\RiskChanges\ComputeLoss.py", line 35, in estimatevulnerability
    vulnerbaility = readIntVuln(con, i)
  File "F:\SDSS\1. code\venv\lib\site-packages\RiskChanges\RiskChangesOps\readvulnerability.py", line 11, in readIntVuln
    vulntable=pd.read_sql(sql_val,engine)        
  File "F:\SDSS\1. code\venv\lib\site-packages\pandas\io\sql.py", line 484, in read_sql
    return pandas_sql.read_query(
  File "F:\SDSS\1. code\venv\lib\site-packages\pandas\io\sql.py", line 1743, in read_query  
    cursor = self.execute(*args)
  File "F:\SDSS\1. code\venv\lib\site-packages\pandas\io\sql.py", line 1709, in execute     
    raise ex from exc
pandas.io.sql.DatabaseError: Execution failed on sql 'SELECT * FROM public."projectIndex_vulvalues" as vulntable WHERE vulntable."vulnID_fk_id"=None;': column "none" does not exist    
LINE 1: ...vulvalues" as vulntable WHERE vulntable."vulnID_fk_id"=None;
```

And below error for the susceptibility hazard type,

```python
Traceback (most recent call last):
  File "F:\SDSS\1. code\venv\lib\site-packages\django\core\handlers\exception.py", line 34, 
in inner
    response = get_response(request)
  File "F:\SDSS\1. code\venv\lib\site-packages\django\core\handlers\base.py", line 115, in _get_response
    response = self.process_exception_by_middleware(e, request)
  File "F:\SDSS\1. code\venv\lib\site-packages\django\core\handlers\base.py", line 113, in _get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "F:\SDSS\1. code\venv\lib\site-packages\django\views\decorators\csrf.py", line 54, in wrapped_view
    return view_func(*args, **kwargs)
  File "F:\SDSS\1. code\venv\lib\site-packages\rest_framework\viewsets.py", line 114, in view
    return self.dispatch(request, *args, **kwargs)
  File "F:\SDSS\1. code\venv\lib\site-packages\rest_framework\views.py", line 505, in dispatch
    response = self.handle_exception(exc)
  File "F:\SDSS\1. code\venv\lib\site-packages\rest_framework\views.py", line 465, in handle_exception
    self.raise_uncaught_exception(exc)
  File "F:\SDSS\1. code\venv\lib\site-packages\rest_framework\views.py", line 476, in raise_uncaught_exception
    raise exc
  File "F:\SDSS\1. code\venv\lib\site-packages\rest_framework\views.py", line 502, in dispatch
    response = handler(request, *args, **kwargs)
  File "F:\SDSS\1. code\django\sdss\loss\views.py", line 100, in create
    loss_calculation(loss_id)
  File "F:\SDSS\1. code\django\sdss\loss\tasks.py", line 26, in loss_calculation
    ComputeLoss(CONN_STR, exposure_id, loss_id, loss_type)
  File "F:\SDSS\1. code\venv\lib\site-packages\RiskChanges\ComputeLoss.py", line 85, in ComputeLoss
    exposure = estimatevulnerability(exposure, haztype, vulnColumn, con)
  File "F:\SDSS\1. code\venv\lib\site-packages\RiskChanges\ComputeLoss.py", line 25, in estimatevulnerability
    subset_exp = pd.merge(left=subset_exp, right=vulnerbaility[[
  File "F:\SDSS\1. code\venv\lib\site-packages\pandas\core\frame.py", line 3030, in __getitem__
    indexer = self.loc._get_listlike_indexer(key, axis=1, raise_missing=True)[1]
  File "F:\SDSS\1. code\venv\lib\site-packages\pandas\core\indexing.py", line 1266, in _get_listlike_indexer
    self._validate_read_indexer(keyarr, indexer, axis, raise_missing=raise_missing)
  File "F:\SDSS\1. code\venv\lib\site-packages\pandas\core\indexing.py", line 1316, in _validate_read_indexer
    raise KeyError(f"{not_found} not in index")
KeyError: "['mean_x'] not in index"
```

**Comment by ashokdahal** [comment]
@iamtekson First error shows:: "vulnID_fk_id"=None;': column "none" does not exist, where None should have been the vulnerability id, could you check if the vulnerability data is linked or not?

**Comment by iamtekson** [comment]
Yes it is linked: 
![image](https://user-images.githubusercontent.com/39838116/139001025-2e5f9232-46b0-482a-9245-4dba0dd669ae.png)

**Comment by ashokdahal** [comment]
@iamtekson  The problem here is, even though some of them are linked, all of the building types are not linked, we analyze each type of buildings by their type and if you see in the above image MA_2 are not linked (null in flash flood) which makes it impossible to calculate the loss. Is it possible to give me the access to Jupiter notebook and also the database so that I can check for the errors myself?

**Comment by iamtekson** [comment]
I think I need to write the request for opening port 8888 to the helpdesk for that. I will write if you need. 

I think the vulnerability connection needs to connect for all the attributes. If there any solution for this kind of partial vuln connection?

**Comment by ashokdahal** [comment]
vulnerability connection needs to connect for all the attributes otherwise loss calculation is not possible because each type of building must have an associated vulnerability curve otherwise if we skip them, the loss calculation will be incomplete making the calculation wrong which is not okay.  Please make the connection to both the database and Jupiter notebook possible so I can check what the problems are. Most of the time problems are occurring due to a mismatch between the data and assumptions in code.

**Comment by iamtekson** [comment]
Ok, sure I will do that. I think it will take 1-2 days to get a response from help-desk.

**Comment by iamtekson** [comment]
I tested the dataset. Now the loss calculation for intensity is working fine for my dataset. The previous error was due to a partial vuln connection. But I am getting the same error for the susceptibility class.

**Comment by ashokdahal** [comment]
Susceptibility is fixed in #50 I am closing this issue
